keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ResourceStatusService.kt

```diff
@@ -103,15 +103,46 @@ class ResourceStatusService(
   }
 
   /**
-   * Checks last 10 events for flapping between only ResourceActuationLaunched and ResourceDeltaDetected
+   * Returns true if a resource has been vetoed by the unhappy veto,
+   * or if the last 10 events are only ResourceActuationLaunched or ResourceDeltaDetected events,
+   * or if the resource has been vetoed by an unspecified veto that we don't have an explicit status mapping for.
    */
   private fun List<ResourceHistoryEvent>.isUnhappy(): Boolean {
+    if (first() is ResourceActuationVetoed && (first() as ResourceActuationVetoed).isUnhappy()) {
+      return true
+    }
+
     val recentSliceOfHistory = this.subList(0, Math.min(10, this.size))
     val filteredHistory = recentSliceOfHistory.filter { it is ResourceDeltaDetected || it is ResourceActuationLaunched }
-    if (filteredHistory.size != recentSliceOfHistory.size) {
-      // there are other events, we're not thrashing.
-      return false
+    if (filteredHistory.size == recentSliceOfHistory.size) {
+      return true
     }
-    return true
+
+    // a catch all for vetoes that don't have an explicit status so they don't get unknown status.
+    return first() is ResourceActuationVetoed
+  }
+
+  /**
+   * Looks at the veto event and determines if it was vetoed by the UnhappyVeto.
+   * Looks at both the [vetoer] and the [reason] for backwards compatibility.
+   */
+  private fun ResourceActuationVetoed.isUnhappy(): Boolean =
+    vetoer?.contains("unhappy", true) ?: false || reason?.contains("unhappy", true) ?: false
+
+  /**
+   * Determines if last event was a veto because of a missing dependency
+   */
+  private fun List<ResourceHistoryEvent>.isMissingDependency(): Boolean =
+    first() is ResourceActuationVetoed && (first() as ResourceActuationVetoed).isMissingDependency()
+
+  /**
+   * Looks at the veto event and determines if it was vetoed by any of the [Required*Veto]s, which indicate a
+   * missing dependency.
+   * Looks at both the [vetoer] and the [reason] for backwards compatibility.
+   */
+  private fun ResourceActuationVetoed.isMissingDependency(): Boolean {
+    val isVetoNew = vetoer?.startsWith("Required", true) ?: false && vetoer?.endsWith("Veto", true) ?: false
+    val isVetoOld = reason?.contains("is not found in", true) ?: false
```

Comment:
this is some deck level backwards compat shenanigans love it ❤️ 🙈

Reply:
You could  add a new `isRequiredDependencyVeto()` function to the `Veto` interface that you check here, instead of relying on the class names.

Reply:
@luispollo I like the idea of adding to the interface, I'll think on that and hopefully improve this!